From: Stefan Bader <stefan.bader@canonical.com>
Date: Wed, 12 Mar 2014 10:51:32 +0100
Subject: libxl: Check contents of capabilities to detect dom0

Only when reading /proc/xen/capabilities returns "control_d" we are running
inside dom0.

BugLink: http://bugs.launchpad.net/bugs/1248025

Signed-off-by: Stefan Bader <stefan.bader@canonical.com>
Index: libvirt-1.1.1/src/libxl/libxl_driver.c
===================================================================
--- libvirt-1.1.1.orig/src/libxl/libxl_driver.c	2014-03-25 14:15:27.810627567 -0500
+++ libvirt-1.1.1/src/libxl/libxl_driver.c	2014-03-25 14:15:27.806627567 -0500
@@ -1196,6 +1196,26 @@ libxlStateInitialize(bool privileged,
         return 0;
     }
 
+    /* Test whether we are running inside a dom0. Testing only for the
+     * presence of the file is not enough as xenfs can be mounted by
+     * any guest.
+     */
+    if (!virFileExists("/proc/xen/capabilities")) {
+        VIR_INFO("Disabling driver as /proc/xen/capabilities does not exist");
+        return 0;
+    }
+    status = virFileReadAll("/proc/xen/capabilities", 10, &output);
+    if (status >= 0) {
+        status = strncmp(output, "control_d", 9);
+    }
+    VIR_FREE(output);
+    output = NULL;
+    if (status) {
+        VIR_INFO("No control_d capabilities detected, probably not running "
+                 "in a Xen Dom0.  Disabling libxenlight driver");
+        return 0;
+    }
+
     /* Disable driver if legacy xen toolstack (xend) is in use */
     cmd = virCommandNewArgList("/usr/lib/xen-common/bin/xen-toolstack", NULL);
     virCommandSetOutputBuffer(cmd, &output);
